x-pack/plugins/task_manager/task_runner.ts

```diff
@@ -0,0 +1,258 @@
+   */
+  public get taskType() {
+    return this.instance.taskType;
+  }
+
+  /**
+   * Gets the task defintion from the dictionary.
+   */
+  public get definition() {
+    return this.definitions[this.taskType];
+  }
+
+  /**
+   * Gets whether or not this task has run longer than its expiration setting allows.
+   */
+  public get isExpired() {
+    return this.instance.runAt < new Date();
+  }
+
+  /**
+   * Returns a log-friendly representation of this task.
+   */
+  public toString() {
+    return `${this.taskType} "${this.id}"`;
+  }
+
+  /**
+   * Runs the task, handling the task result, errors, etc, rescheduling if need
+   * be. NOTE: the time of applying the middleware's beforeRun is incorporated
+   * into the total timeout time the task in configured with. We may decide to
+   * start the timer after beforeRun resolves
+   *
+   * @returns {Promise<RunResult>}
+   */
+  public async run(): Promise<RunResult> {
+    try {
+      this.logger.debug(`Running task ${this}`);
+      const modifiedContext = await this.beforeRun({
+        kbnServer: this.kbnServer,
+        taskInstance: this.instance,
+      });
+      const task = this.definition.createTaskRunner(modifiedContext);
+      this.task = task;
+      return this.processResult(this.validateResult(await this.task.run()));
+    } catch (error) {
+      this.logger.warning(`Task ${this} failed ${error.stack}`);
+      this.logger.debug(`Task ${JSON.stringify(this.instance)} failed ${error.stack}`);
+
+      return this.processResult({ error });
+    }
+  }
+
+  /**
+   * Attempts to claim exclusive rights to run the task. If the attempt fails
+   * with a 409 (http conflict), we assume another Kibana instance beat us to the punch.
+   *
+   * @returns {Promise<boolean>}
+   */
+  public async claimOwnership(): Promise<boolean> {
+    const VERSION_CONFLICT_STATUS = 409;
+
+    try {
+      this.instance = await this.store.update({
+        ...this.instance,
+        status: 'running',
+        runAt: intervalFromNow(this.definition.timeOut)!,
+      });
+
+      return true;
+    } catch (error) {
+      if (error.statusCode !== VERSION_CONFLICT_STATUS) {
+        throw error;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Attempts to cancel the task.
+   *
+   * @returns {Promise<void>}
+   */
+  public async cancel() {
+    const { task } = this;
+    if (task && task.cancel) {
+      this.task = undefined;
+      return task.cancel();
+    }
+
+    this.logger.warning(`The task ${this} is not cancellable.`);
+  }
+
+  private validateResult(result?: RunResult | void): RunResult {
+    const { error } = Joi.validate(result, validateRunResult);
+
+    if (error) {
+      this.logger.warning(`Invalid task result for ${this}: ${error.message}`);
+    }
+
+    return result || {};
+  }
+
+  private async processResult(result: RunResult): Promise<RunResult> {
+    const recurring = result.runAt || this.instance.interval || result.error;
+    if (recurring) {
```

Comment:
Given the number of comments I've added to this function, it may help to split this function into `processResultForRecurringTask` and `processResultWhenDone`, then simply doing:

```ts
if (result.runAt || this.instance.interval || result.error) {
  await this.processResultForRecurringTask(result);
} else {
  await this.processResultWhenDone(result);
}

return result;
```

Reply:
1ffdf38828